### PR TITLE
Support for JaCoCo on Java 9

### PR DESCRIPTION
**Pull request by bmuschko**
For more information see #1006.

### CODE REVIEW

**Comment by bmuschko** [comment]
Relevant [CI job](https://builds.gradle.org/viewLog.html?buildId=1976250&tab=buildResultsDiv&buildTypeId=Gradle_Branches_Checkpoints_Stage4FullCoverage).

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy [hunk 1]
```diff
@@ -43,7 +43,7 @@ class JacocoVersionIntegTest extends MultiVersionIntegrationSpec {
         then:
         def report = htmlReport()
         report.totalCoverage() == 100
-        report.jacocoVersion() == version
+        report.jacocoVersion() == version || report.jacocoVersion().startsWith(version)
```
Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy]: Can this test be made to not require <=JDK7?

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/rules/JacocoPluginCoverageVerificationIntegrationTest.groovy [hunk 2]
```diff
@@ -18,14 +18,11 @@ package org.gradle.testing.jacoco.plugins.rules
...
 import org.gradle.testing.jacoco.plugins.fixtures.JavaProjectUnderTest
-import org.gradle.util.Requires
 import spock.lang.Unroll
 
 import static JacocoViolationRulesLimit.Insufficient
 import static JacocoViolationRulesLimit.Sufficient
-import static org.gradle.util.TestPrecondition.FIX_TO_WORK_ON_JAVA9
 
-@Requires(FIX_TO_WORK_ON_JAVA9)
 class JacocoPluginCoverageVerificationIntegrationTest extends AbstractIntegrationSpec {
```
Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/rules/JacocoPluginCoverageVerificationIntegrationTest.groovy]: can this be made into a `MultiVersionIntegrationSpec` too?

**Comment by big-guy** [comment]
The other thing I forgot to mention -- we should add something to the release notes about a default version change.

**Comment by bmuschko** [comment]
@big-guy Yeah, I was planning to add something to the release notes after merging.

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy (hunk 1 above)
Comment (bmuschko) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy]: Unfortunately no as older versions of JaCoCo don't work on JDK 8.

Comment (bmuschko) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy]: The latest JaCoCo library version has a non-timestamped version but they still use the timestamp when generating the report. Doh! I think it's simply an oversight on their end.

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy [hunk 3]
```diff
@@ -267,19 +257,17 @@ public class ThingTest {
         ":test" in nonSkippedTasks
         ":otherTests" in nonSkippedTasks
         file("build/jacoco/jacocoMerge.exec").exists()
-        htmlReport("build/reports/jacoco/mergedReport/html").totalCoverage() == 65
+        htmlReport("build/reports/jacoco/mergedReport/html").totalCoverage() == 64
```
Comment (bmuschko) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy]: I think something changes on the JaCoCo side on how they calculate the coverage number. I didn't look into their release notes though. Maybe they had a bug before or introduced a new one.

Comment (bmuschko) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy]: Here's the reason: https://github.com/jacoco/jacoco/pull/452

Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy]: Thanks, could we change our sample project to make this the same across versions?

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy (hunk 1 above)
Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoVersionIntegTest.groovy]: Can we move that detail into the report fixture? e.g., so we can do something like:

`report.assertJacocoVersion(version)`

And `assertJacocoVersion` does the fuzzy match.

**Comment by bmuschko** [comment]
@big-guy I addressed your comments for this PR. Can you please have another look?

**Pull request merged**